PROJ-133 : Update section 5.2 Receiving the information (#70)
SW805F20/report: sections/sprint-4/client-side-handling.tex

\subsection*{UDP client}
Handling the datagrams on the UDP client is simple as it can currently only receive a message with type 0, which is to a message containing a position for a Pozyx tag.

> The datagram is recieved in the format \textit{0xYYYYXXXXIISSTT} where:

Reviewer: ```suggestion
The datagram is received in the format \textit{0xYYYYXXXXIISSTT} where 0x indicates hex and the rest of the identifiers are:
```

> The datagram is received in the format \textit{0xYYYYXXXXIISSTT} where 0x indicates hex and the rest of the identifiers are:

Reviewer: texttt

\begin{itemize}
    \item Y - Y position of tag
    \item X - X position of tag
    \item I - Id of player
    \item S - Timestamp

>     \item T - Package type

Reviewer: : i stedet for -

> \end{itemize}

Reviewer: kunne være nice som tabel

> When this datagram is received, we start by parsing the hexadecimal to a long.

Reviewer: ```suggestion
When this datagram is received, we start by parsing the hexadecimal to a long datatype.
```

> Because one hexadecimal corresponds to four bits we can get type of the datagram which is the last two hexadecimals of the datagram by typecasting it to a byte.

Reviewer: ```suggestion
Since one hexadecimal corresponds to four bits we can get the type of the message, which is the last two hex values of the datagram by typecasting it to a byte.
```

> As seen on line 13 in \autoref{lst:readingudpdatagram}.

Reviewer: Det her skal være med i ovenstående linje eller om formuleres

    byte type;
    if (long.TryParse(datagramMessage, System.Globalization.NumberStyles.HexNumber, System.Globalization.CultureInfo.InvariantCulture, out data))
    {

>         type = (byte)data;|\label{line:decode-type}|

Reviewer: ?????

![image](https://user-images.githubusercontent.com/21969475/80682750-5d7dc200-8ac3-11ea-84a3-c309987c3b4b.png)

    
}
\end{lstlisting}
\noindent 
The next step in processing the data is to read the actual content of the message sent.
This is done by utilizing the fact that all parts of the message have a size corresponding to a type in C\#.
The datagram is read from right to left using the right bitshift operation.
When a hexadecimal is right bit shifted 4 bits every decimal in the hexadecimal is moved one decimal to the right.
\begin{center}
    0x123 >> 4 -> 0x012   
\end{center}
This can be utilized to read the different parts of the message that the client received.
The first part that is read from the message is timestamp.
In the message the decimals representing the timestamp are followed by two decimals representing the type of the message as seen earlier.
This means that because one decimal corresponds to four bits, the message has to be right bit shifted by 8 bits to move the timestamp to the back of the hexadecimal.
\begin{center}

>     0xYYYYXXXXIISSTT >> 8 -> 0x00YYYYXXXXIISS

Reviewer: det her kunne være nicere med en grafisk illustration, det er lettere at overskue

evt bruge en pil der peger på hvor vi har bitshifted hen

}
\end{lstlisting}

> The next step in processing the data is to read the actual content of the datagram sent, in this step we utilize that the fact that all messages have a size corresponding to a type in C\#.

Reviewer: Vi kalder det messages/packets i network delen?

We continue to read the datagram from right to left using the right bitshift operation.
The first part we want to read from the message is the timestamp, which we know from the network format (\autoref{app:network}) has a size of 8 bits, which corresponds to the byte type in C\#.
We know that the type of the package is 8 bits, we can therefore remove the type and get the timestamp to the back of the datagram by right bitshifting it by 8 bits as can be seen on line 5 in \autoref{lst:updateplayerdata} 

> We continue doing this for each message in the datagram where we start by right shifting the datagram by the combined size of the previously read messages and then typecasting it to a variable type with the same size as the message that is being read.

Reviewer: Synes at det her er lidt rodet, og der står godt nok we mange gange

Reviewer: Og er ikke sikker på hvor meget det forklarer det egentlig

Reviewer: Måske en lille illustration egentligf?????
